File: src/core/moves/abilities/Destroy.ts
import { destroyStructure } from "../../actions";
import { CallbackResult } from "../../move";
import { GameState } from "../../states";
import { AbilityType } from "../../types";
import Ability from "../Ability";

export default class Destroy extends Ability {
    constructor(target: number) {
        super(null, target, AbilityType.Destroy);
    }

    execute(state: GameState): CallbackResult {
        return {
            rewards: [],
            undo: destroyStructure(state, this.getTarget()),
        };
    }
}

File: polyfish-rs/src/moves/abilities/destroy.rs
use crate::moves::{Move, MoveResult};
use crate::states::GameState;
use crate::types::{AbilityType, MoveType};

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DestroyMove {
    pub target_index: i32,
}

impl DestroyMove {
    pub fn new(target_index: i32) -> Self {
        Self { target_index }
    }
}

impl Move for DestroyMove {
    fn move_type(&self) -> MoveType {
        MoveType::Ability
    }

    fn execute(&self, state: &mut GameState) -> Result<MoveResult, String> {
        let undo = crate::actions::structure::destroy_structure(state, self.target_index);
        Ok(MoveResult {
            undo,
            rewards: None,
        })
    }

    fn describe(&self, _state: &GameState) -> String {
        format!("Destroy structure at {}", self.target_index)
    }

    fn serialize(&self) -> serde_json::Value {
        serde_json::json!({
            "moveType": self.move_type(),
            "type": self.ability_type(),
            "target": self.target_index,
        })
    }

    #[inline]
    fn target_idx(&self) -> Result<usize, String> {
        Ok(self.target_index as usize)
    }

    #[inline]
    fn ability_type(&self) -> Result<AbilityType, String> {
        Ok(AbilityType::Destroy)
    }
}
